noir_stdlib/src/meta/expr.nr

```diff
@@ -110,4 +110,51 @@ impl Expr {
     // docs:start:is_continue
     fn is_continue(self) -> bool {}
     // docs:end:is_continue
+
+    fn map<Env>(self, f: fn[Env](Expr) -> Option<Expr>) -> Expr {
+        let result = f(self);
+        let result = result.or_else(|| map_if(self, f));
+        let result = result.or_else(|| map_block(self, f));
+        // TODO: check the remaining possibilities...
+        // though to test this it's easier if we have `as_let`.
+        result.unwrap_or(self)
+    }
+}
+
+fn map_block<Env>(expr: Expr, f: fn[Env](Expr) -> Option<Expr>) -> Option<Expr> {
+    let exprs = expr.as_block();
+    if exprs.is_some() {
+        let exprs = exprs.unwrap();
+        let exprs = exprs.map(|expr: Expr| expr.map(f));
+        Option::some(new_block(exprs))
+    } else {
+        Option::none()
+    }
```

Comment:
```suggestion
    expr.as_block().map(|exprs| {
        let exprs = exprs.map(|expr: Expr| expr.map(f));
        new_block(exprs)
    })
```
Can do something similar for other methods.

You can't do the following for `if` since it has an optional else, but for many other variants you could feed them to one function `map_helper(exprs: [Expr], f) -> [Expr]` which takes a slice so you can just bundle up differing number of exprs and map them all with that. You'd still have to unpack the returned slice manually though.